### PR TITLE
auto master-slave switch

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve? 
- [x] syncer auto switch relay subdir when master-slave switch
- [x] auto test master-slave switch
- [x] auto detect master-slave switch

Tests 

 - Unit test
 - github action

### CODE REVIEW

File: pkg/streamer/file_test.go [hunk 1]
```diff
@@ -379,8 +394,10 @@ func (t *testUtilSuite) TestRelaySubDirUpdated(c *C) {
...
 	go func() {
 		defer wg.Done()
-		up, err2 := relaySubDirUpdated(ctx, watcherInterval, subDir, relayPaths[1], relayFiles[1], 0)
-		c.Assert(err2, IsNil)
+		relaySubDirUpdated(ctx, watcherInterval, subDir, relayPaths[1], relayFiles[1], 0, updatePathCh, errCh)
+		c.Assert(len(errCh), Equals, 0)
+		c.Assert(len(updatePathCh), Equals, 1)
+		up := <-updatePathCh
 		c.Assert(up, Equals, relayPaths[1])
 	}()
```
Comment (lance6716) [inline pkg/streamer/file_test.go:403]: in line 404 there's `// wait watcher started, update the file`, don't know why this is needed

File: pkg/streamer/file_test.go [hunk 2]
```diff
@@ -412,8 +431,10 @@ func (t *testUtilSuite) TestRelaySubDirUpdated(c *C) {
 	wg.Add(1)
```
Comment (lance6716) [inline pkg/streamer/file_test.go:431]: seems line 442 to 460 should be put above this line

**Review by lance6716 — COMMENTED** [review]
will review soon

File: relay/relay.go [hunk 3]
```diff
@@ -451,6 +451,19 @@ func (r *Relay) handleEvents(ctx context.Context, reader2 reader.Reader, transfo
...
 		}
 
+		// fake rotate event
+		if _, ok := e.Event.(*replication.RotateEvent); ok && e.Header.Timestamp == 0 && e.Header.LogPos == 0 {
+			isNew, err2 := isNewServer(ctx, r.meta.UUID(), r.db, r.cfg.Flavor)
+			if err2 != nil {
+				return err2
+			}
+			// upstream database switch
+			// report an error, let outer logic handle it
```
Comment (lance6716) [inline relay/relay.go:461]: where is handling logic 😵

Comment (GMHDBJD) [reply relay/relay.go:461]: now it report an error，we handle it by auto resume or user resume it manually

Comment (GMHDBJD) [reply relay/relay.go:461]: handle in https://github.com/pingcap/dm/blob/a778df270b62dd90adcc513f57469e8b331748ab/relay/relay.go#L201

**Comment by lance6716** [comment]
PTAL @3pointer @amyangfei

File: pkg/streamer/file.go [hunk 4]
```diff
@@ -269,60 +279,91 @@ func relaySubDirUpdated(ctx context.Context, watcherInterval time.Duration, dir
...
-		// because creating sub directory and writing relay log file are not atomic
-		// so we let user to pause syncing before switching relay's master server
-		return false, false, "", "", err
-	}
+			// try get the first binlog file in next sub directory
+			nextBinlogName, err = getFirstBinlogName(relayDir, nextUUID)
+			if err != nil {
+				// because creating sub directory and writing relay log file are not atomic
+				// just continue to observe subdir
+				if terror.ErrBinlogFilesNotFound.Equal(err) {
```
[the remark this answers is not in the document]
Comment (GMHDBJD) [reply pkg/streamer/file.go:358]: If terminated with an empty folder, relay will continue from that empty folder if index file updated or previous folder if index file not updated. See https://github.com/pingcap/dm/blob/a778df270b62dd90adcc513f57469e8b331748ab/relay/relay.go#L159

File: pkg/streamer/file_test.go (hunk 1 above)
Comment (GMHDBJD) [reply pkg/streamer/file_test.go:403]: Watcher only catch update event after watch. See https://github.com/pingcap/dm/blob/master/pkg/streamer/file.go#L198-L203

File: pkg/streamer/file_test.go (hunk 2 above)
Comment (GMHDBJD) [reply pkg/streamer/file_test.go:431]: ditto

File: pkg/streamer/reader.go [hunk 5]
```diff
@@ -605,29 +605,61 @@ func (r *BinlogReader) parseFile(
...
-	if err != nil {
-		return false, false, 0, "", "", err
-	} else if needReParse {
-		// need to re-parse the current relay log file
-		return false, true, latestPos, "", "", nil
-	} else if needSwitch {
-		// need to switch to next relay sub directory
-		return true, false, 0, nextUUID, nextBinlogName, nil
-	}
+	switchCh := make(chan struct {
```
[the remark this answers is not in the document]
Comment (GMHDBJD) [reply pkg/streamer/reader.go]: address in 2530587

File: pkg/streamer/reader.go [hunk 6]
```diff
@@ -605,29 +605,61 @@ func (r *BinlogReader) parseFile(
...
+			return false, false, 0, "", "", nil
+		}
+		return true, false, 0, switchResp.nextUUID, switchResp.nextBinlogName, nil
+	case updatePath := <-updatePathCh:
+		if strings.HasSuffix(updatePath, relayLogFile) {
+			// current relay log file updated, need to re-parse it
+			return false, true, latestPos, "", "", nil
+		}
+		// need parse next relay log file or re-collect files
+		return false, false, latestPos, "", "", nil
```
[the remark this answers is not in the document]
Comment (GMHDBJD) [reply pkg/streamer/reader.go:654]: Seems it only occur in this func. I keep it here.

File: pkg/streamer/reader.go [hunk 7]
```diff
@@ -605,29 +605,58 @@ func (r *BinlogReader) parseFile(
...
+			close(updateErrCh)
+			wg.Done()
+		}()
+		relaySubDirUpdated(newCtx, watcherInterval, relayLogDir, fullPath, relayLogFile, latestPos, updatePathCh, updateErrCh)
+	}(latestPos)
+
+	select {
+	case <-ctx.Done():
+		return false, false, 0, "", "", nil
+	case switchResp := <-switchCh:
```
Comment (lance6716) [inline pkg/streamer/reader.go:642]: will relay break-then-autoresume, so generated a new subdirectory. meanwhile old subdirectory has updated something before break, so in this `select` can't assure priority?

**Review by lance6716 — COMMENTED** [review]
rest LGTM

File: pkg/streamer/reader.go (hunk 7 above)
Comment (GMHDBJD) [reply pkg/streamer/reader.go:642]: add a check in 20f40cd

**Comment by lance6716** [comment]
LGTM

**Comment by lance6716** [comment]
PTAL @3pointer @amyangfei

**Review by lichunzhu — COMMENTED** [review]
Rest LGTM

**Comment by lance6716** [comment]
Sorry I commited https://github.com/pingcap/dm/pull/1364/commits/76ddfedf59d1e2e34686c51c65d0ccbcc6b4915c, but it need `make terror_check` @GMHDBJD 😂

File: errors.toml [hunk 8]
```diff
@@ -1390,6 +1390,12 @@ description = ""
 workaround = ""
 tags = ["internal", "high"]
 
+[error.DM-relay-unit-30044]
+message = "receive fake rotate event with differend server_id"
```
Comment (lance6716) [inline errors.toml:1394]:
```suggestion
message = "receive fake rotate event with different server_id"
```

File: _utils/terror_gen/errors_release.txt [hunk 9]
```diff
@@ -230,6 +230,7 @@ ErrRelayTrimUUIDNotFound,[code=30040:class=relay-unit:scope=internal:level=high]
 ErrRelayRemoveFileFail,[code=30041:class=relay-unit:scope=internal:level=high], "Message: remove relay log %s %s"
 ErrRelayPurgeArgsNotValid,[code=30042:class=relay-unit:scope=internal:level=high], "Message: args (%T) %+v not valid"
 ErrPreviousGTIDsNotValid,[code=30043:class=relay-unit:scope=internal:level=high], "Message: previousGTIDs %s not valid"
+ErrRotateEventWithDifferentServerID,[code=30044:class=relay-unit:scope=internal:level=high], "Message: receive fake rotate event with differend server_id, Workaround: Please use `resume-relay` command if upstream database has changed"
```
Comment (lance6716) [inline _utils/terror_gen/errors_release.txt:233]:
```suggestion
ErrRotateEventWithDifferentServerID,[code=30044:class=relay-unit:scope=internal:level=high], "Message: receive fake rotate event with different server_id, Workaround: Please use `resume-relay` command if upstream database has changed"
```

**Comment by lance6716** [comment]
required CI passed, PTAL @lichunzhu

**Review by lichunzhu — APPROVED** [review]
LGTM

**Pull request merged**